io-engine/src/rebuild/rebuild_log.rs

```diff
@@ -0,0 +1,158 @@
+use bit_vec::BitVec;
+use parking_lot::Mutex;
+use spdk_rs::IoType;
+use std::{
+    fmt::{Debug, Formatter},
+    sync::Arc,
+};
+
+use super::SEGMENT_SIZE;
+
+// Returns ceil of an integer division.
+fn div_ceil(a: u64, b: u64) -> u64 {
+    (a + b - 1) / b
+}
+
+/// Rebuild log data structure.
+pub struct RebuildLog {
+    /// Name of the underlying device, stored for diagnostic purposes.
+    device_name: String,
+    /// Segment bitmap. Ones indicate modified segments.
+    segments: BitVec,
```

Comment:
Hi! @dsavitskiy I Hope you are well!

I was wondering if you have considered using a bitmap of dirty blocks rather than dirty segments as it avoids calculations for each write operation. When you want to replay the log or whatever, you can "pay the price" of blocks to segments then?

Reply:
Hi Jeffry!

Bitmap of dirty blocks (512 bytes) would be too large, and our rebuild is running at "segment" granularity anyway.
Calculating block to segment is just an integer multiplication and an integer division. Our code path is not that much computation bound.
And the faulty child is out of I/O path anyway, so we have few spare CPU cycles.
Again, bloating bitmap to avoid a division may be slower, not faster, because of more cache misses. Waiting for a line to fetch is certainly more than a int division.